### DISCUSSION TITLE
[A-L] (12 gennaio 2023) Esercizio 1

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#71] [Esami A-L]
_tengo a precisare che sulla traccia originale ha scritto 2016, si è sbagliato, è del 2023._
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/bf029d37-fb53-43db-bdb9-5aa9bd8f9767)

**Tradotto dal Wollanese**: Il minimo comunue antenato di $x,y,z$ è il nodo più lontano dalla radice che è genitore/nonno di $x,y,z$

_(grazie @faciolox per la correzione, se ho risbagliato correggimi)_

**Comment by faciolox** [comment]
Il minimo comune antenato in realtá dovrebbe essere l'antenato comune (quindi il nodo che ha discendente sia x che y che z) che ha il livello piú basso dell'albero.

Qua uso una funzione che trova il path dal nodo alla radice che ha tempo O(n).

Poi uso una funzione che trova il minimo antenato tra due nodi. Trovo i path per la radice dei due nodi e li confronto fino a che l'antenato non cambia. L'ultimo in comune sara il minimo antenato comune.

in un altra funzione trovo il minimo antenato comune tra x e y e poi lo confronto con z trovando il minimo tra i tre

![Screenshot 2024-06-17 alle 11 04 27](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/99180436/b11d4450-748b-417a-8df5-77efcd8f2c39)

**Comment by IonutCicio** [comment]
Possiamo salvarci in due array di booleani quali sono gli antenati di `x` e `y`.
A questo punto, ripercorrendo gli antenati di `z`, se troviano un nodo che è sia antenato di `x` sia antenato di `y`, abbiamo trovato il primo antenato comune.

```rust
fn triple_first_common_daddy(tree: &[usize], mut x: usize, mut y: usize, mut z: usize) -> usize {
    let mut x_daddies = vec![false; tree.len()];
    let mut y_daddies = vec![false; tree.len()];

    while x != tree[x] {
        x = tree[x];
        x_daddies[x] = true;
    }

    while y != tree[y] {
        y = tree[y];
        y_daddies[y] = true;
    }

    while z != tree[z] {
        z = tree[z];
        if x_daddies[z] && y_daddies[z] {
            return z;
        }
    }

    x
}
```

**Comment by cornflexxx** [comment]
Consideriamo le distanze dalla radice di  $x,y,z$, rispettivamente $d_x,d_y,d_z$ calcolate in tempo $O(n)$, per trovare l'antenato minimo comune è sufficiente risalire a partire dai tre vertici l'albero finché $d_x=d_y=d_z$. 
Successivamente possiamo risalire insieme tutti e tre i vertici $x,y,z$ arrivando nello stesso momento all'antenato comune.

**Comment by daromacarlo** [comment]
Il costo è O(n) dato che ogni coclo presente nel codice ha costo O(n)
```py
def antenati_aux(P, x, y, z):
    h = x
    k = z
    j = y
 
    lista_x = [0] * len(P)
    lista_y = [0] * len(P)
    lista_z = [0] * len(P)
 
    count = 0
    while P[h] != h:
        lista_x[h] = count + 1
        h = P[h]
    
    count = 0
    while P[j] != j:
        lista_y[j] = count + 1
        j = P[j]
    
    count = 0
    while P[k] != k:
        lista_z[k] = count + 1
        k = P[k]
    
    # Determina l'antenato comune più vicino
    min_value = float('inf')
    for m in range(len(P)):
        if lista_x[m] >= 1 and lista_y[m] >= 1 and lista_z[m] >= 1:
            total = lista_x[m] + lista_y[m] + lista_z[m]
            if total < min_value:
                min_value = total
    
    return min_value

```
